Validate requests using swagger spec (#1465)

amarjandu opened the pull request:
Author

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue

Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>

Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>

Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

Primary reviewer (after approval)

- [ ] Commented on demo expectations                        <sub>or labelled PR as `no demo`</sub>
- [ ] Decided if PR can be labeled `no sandbox`
- [ ] PR title is appropriate as title of merge commit
- [ ] Moved ticket to Approved column
- [ ] Assigned PR to an operator

Operator (before pushing merge the commit)

- [ ] Checked `reindex` label and `r` commit title tag
- [ ] Checked that Demo expectations are clear              <sub>or PR is labeled as `no demo`</sub>
- [ ] Rebased and squashed branch
- [ ] Sanity-checked history
- [ ] Pushed PR branch to Github
- [ ] Branch pushed to Gitlab and build passes in `sandbox` <sub>or added `no sandbox` label</sub>
- [ ] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [ ] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [ ] Added PR reference to merge commit title
- [ ] Collected commit title tags in merge commit title
- [ ] Moved linked issue to Merged column
- [ ] Pushed merge commit to Github

Operator (after pushing the merge commit)

- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab

Operator (reindex) 

- [ ] Purchased BQ slot committment in `dev`                <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Started reindex in `dev`                              <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Deleted BQ slot committment in `dev`                  <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Checked for failures in `dev`                         <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Purchased BQ slot committment in `prod`               <sub>or this PR does not require reindexing `prod`</sub>
- [ ] Started reindex in `prod`                             <sub>or this PR does not require reindexing `prod`</sub>
- [ ] Deleted BQ slot committment in `prod`                 <sub>or this PR does not require reindexing `prod`</sub>
- [ ] Checked for failures in `prod`                        <sub>or this PR does not require reindexing `prod`</sub>

Operator

- [ ] Unassigned PR

amarjandu committed:
[R 2/2] Validate requests using Swagger spec (#1465)

Add new validation functionality to ValidateAzulChaliceApp.
Use ValidateAzulChaliceApp as parent class for ServiceApp.
Allow for deterministic ordering for 400 responses.

amarjandu commented:
Waiting on full documentation for `api_specs`. 

TODO: 
1. resolve test failures when those land. 
1. maybe make make it feel like its less `in-band`

achave11-ucsc reviewed (changes requested):
It looks great overall. 
There was just one [observation](https://github.com/DataBiosphere/azul/pull/1849#discussion_r432973852) regarding the validation in `app.py`

The refactoring should be done in a separate commit, so that we could distinguish with ease what changes you introduced and what was already present.

amarjandu committed:
[R 1/2] Validate requests using Swagger spec (#1465)

Remove old validation methods used in Service app.

danielsotirhos commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,183 @@
+
+    spec_parameters = build_spec_validation(api_specs, resource_path, http_method)
+    for name, spec_parameter in spec_parameters.items():
+        if spec_parameter.required:
```
PyCharm complains this is an `Unresolved attribute reference 'required' for class 'SpecValidation'`.
I see `SpecValidation` objects are created with dynamic arguments (`temp = SpecValidation(**parameter)`) however I feel since `required` is expected to always be defined it should be declared in the class.

danielsotirhos commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,183 @@
+        spec_parameters[param_name].validate(dict(name=param_name, value=param_value))
+
+
+def _extract_parameters(api_spec: JSON, resource_path: list = None, http_method: str = None) -> Mapping:
```
```suggestion
def _extract_parameters(api_spec: JSON, resource_path: str = None, http_method: str = None) -> List[JSON]:
```

danielsotirhos commented on test/service/test_request_validation.py:
```diff
@@ -118,7 +124,8 @@ def test_bad_sort_facet_but_valid_filter_facet_of_sample(self):
         response = requests.get(url, params=params)
         self.assertEqual(400, response.status_code, response.json())
-        self.assertEqual(self.facet_message, response.json())
+        self.assertEqual(self.facet_enumeration_message(),
+                         response.json())
```
```suggestion
        self.assertEqual(self.facet_enumeration_message(), response.json())
```
A single line seems fine here

danielsotirhos commented on test/service/test_request_validation.py:
```diff
@@ -163,7 +170,8 @@ def test_bad_sort_facet_of_file(self):
         response = requests.get(url, params=params)
         self.assertEqual(400, response.status_code, response.json())
-        self.assertEqual(self.facet_message, response.json())
+        self.assertEqual(self.facet_enumeration_message(),
+                         response.json())
```
```suggestion
        self.assertEqual(self.facet_enumeration_message(), response.json())
```

danielsotirhos commented on test/service/test_request_validation.py:
```diff
@@ -186,7 +194,8 @@ def test_bad_sort_facet_but_valid_filter_facet_of_file(self):
         response = requests.get(url, params=params)
         self.assertEqual(400, response.status_code, response.json())
-        self.assertEqual(self.facet_message, response.json())
+        self.assertEqual(self.facet_enumeration_message(),
+                         response.json())
```
```suggestion
        self.assertEqual(self.facet_enumeration_message(), response.json())
```

danielsotirhos commented on test/test_app_logging.py:
```diff
@@ -73,7 +73,8 @@ def fail():
                     # The exception is always logged
                     self.assertEqual(len(app_log.output), 1)
-                    err_log = 'ERROR:test_app_logging:Caught exception for <function TestAppLogging.test.<locals>.fail'
+                    err_log = 'ERROR:test_app_logging:Caught exception for ' \
+                              '<function AzulChaliceApp.route.<locals>.decorator.<locals>.f'
```
```suggestion
                    err_log = 'ERROR:test_app_logging:Caught exception for ' \
                              '<function AzulChaliceApp.route.<locals>.decorator.<locals>.fail'
```

amarjandu replied:
`f` is from the method inside the decorator in `src/azul/Chalice.py` ln 82.

dsotirho-ucsc reviewed (changes requested):
Looking great. Just a few cleanup items.

amarjandu commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,183 @@
+        def _filter_validate(request_parameter: Mapping):
+            spec_schema = self.content['application/json']['schema']
+            try:
+                filter = json.loads(request_parameter['value'])
```
good catch

danielsotirhos commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,187 @@
+        :param request_parameter: A query parameter dict with 'name' and 'value' keys
+        """
+
+        def _schema_validate(request_parameter: Mapping):
```
```suggestion
        def _schema_validate(request_parameter: Mapping[str, str]):
```

danielsotirhos commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,187 @@
+    List,
+    Mapping,
+)
+from chalice import (
```
One blank line needed between first group of Python-bundled modules and second group of external modules.

danielsotirhos commented on test/service/test_request_validation.py:
```diff
@@ -81,7 +86,8 @@ def test_bad_sort_facet_of_sample(self):
         response = requests.get(url, params=params)
         self.assertEqual(400, response.status_code, response.json())
-        self.assertEqual(self.facet_message, response.json())
+        self.assertEqual(self.facet_enumeration_message(),
+                         response.json())
```
```suggestion
        self.assertEqual(self.facet_enumeration_message(), response.json())
```

dsotirho-ucsc reviewed (changes requested):
Sorry, just a couple more cleanup items that I missed the first time.

danielsotirhos commented on test/service/test_request_validation.py:
```diff
@@ -186,8 +185,8 @@ def test_bad_sort_facet_but_valid_filter_facet_of_file(self):
             'size': 15,
-            'sort': 'bad-facet',
-            'order': 'asc',
+            'sort': 'bad-facet', ''
+                                 'order': 'asc',
```
Typo?

danielsotirhos commented on test/service/test_request_validation.py:
```diff
@@ -180,13 +185,13 @@ def test_bad_sort_facet_but_valid_filter_facet_of_file(self):
             'size': 15,
-            'sort': 'bad-facet',
-            'order': 'asc',
+            'sort': 'bad-facet', ''
+                                 'order': 'asc',
```
Typo?

amarjandu replied:
yeah not sure how its formated so nicely. waiting on tests my bad.

dsotirho-ucsc reviewed (approved):
Looking good. Approved.

danielsotirhos commented on test/service/test_request_validation.py:
```diff
@@ -284,10 +296,36 @@ def test_bad_filter_relation(self):
+            path_specs = _extract_parameters(api_spec=self.app_module.app.specs,
+                                             resource_path=path,
+                                             http_method='get')
+            filter_spec = next(x for x in path_specs if x['name'] == 'filters')
+            spec_location = ['content', 'application/json', 'schema', 'properties', facet, 'oneOf']
+            relation_spec = next(x for x in reduce(getitem, spec_location, filter_spec) if relation in x['properties'])
```
Using `more_itertools.one` would make this stronger.
```suggestion
            filter_spec = one(x for x in path_specs if x['name'] == 'filters')
            spec_location = ['content', 'application/json', 'schema', 'properties', facet, 'oneOf']
            relation_spec = one(x for x in reduce(getitem, spec_location, filter_spec) if relation in x['properties'])
```

dsotirho-ucsc reviewed (approved):
Looks good for approval as is. Found just a couple things to consider updating.

hannes-ucsc commented on lambdas/service/app.py:
```diff
@@ -306,7 +300,8 @@ def __init__(self):
                          unit_test=globals().get('unit_test', False),
-                         spec=spec)
+                         spec=spec,
+                         validation_method=validation.validate_params)
```
`ValidatingAzulChaliceApp`

hannes-ucsc commented on lambdas/service/app.py:
```diff
@@ -2052,7 +1816,8 @@ def add_item_to_cart(cart_id):
 
 
-@app.test_route('/resources/carts/{cart_id}/items/{item_id}', methods=['DELETE'], cors=True, authorizer=jwt_auth)
+@app.test_route('/resources/carts/{cart_id}/items/{item_id}', methods=['DELETE'], cors=True, authorizer=jwt_auth,
```
wrap all or none

hannes-ucsc commented on src/azul/chalice.py:
```diff
@@ -30,8 +30,9 @@
 class AzulChaliceApp(Chalice):
 
-    def __init__(self, app_name, unit_test=False, spec=None):
+    def __init__(self, app_name, unit_test=False, spec=None, validation_method=None):
```
see above

hannes-ucsc commented on src/azul/chalice.py:
```diff
@@ -69,17 +71,25 @@ def route(self,
                 # the function itself.
-                view_func.path = path
-                return chalice_decorator(view_func)
+                def f(*args, **kwargs):
```
functools.wraps

hannes-ucsc commented on src/azul/openapi/schema.py:
```diff
@@ -330,13 +330,19 @@ def in_range(minimum: Optional[N], maximum: Optional[N], type_: Optional[TYPE] =
-    str: {'type': 'string'},
     bool: {'type': 'boolean'},
-    int: {'type': 'integer', 'format': 'int64'},
+    dict: {'type': 'object'},
```
dict is not a primitive type

amarjandu replied:
refactored

hannes-ucsc commented on src/azul/openapi/schema.py:
```diff
@@ -330,13 +330,19 @@ def in_range(minimum: Optional[N], maximum: Optional[N], type_: Optional[TYPE] =
+    dict: {'type': 'object'},
     float: {'type': 'number', 'format': 'double'},
+    int: {'type': 'integer', 'format': 'int64'},
+    list: {'type': 'array'},
```
neither is list.

amarjandu replied:
refactored

hannes-ucsc commented on src/azul/openapi/schema.py:
```diff
@@ -330,13 +330,19 @@ def in_range(minimum: Optional[N], maximum: Optional[N], type_: Optional[TYPE] =
-    int: {'type': 'integer', 'format': 'int64'},
+    dict: {'type': 'object'},
     float: {'type': 'number', 'format': 'double'},
+    int: {'type': 'integer', 'format': 'int64'},
```
minimize diffs

amarjandu replied:
reverted ordering back to develop

hannes-ucsc commented on src/azul/openapi/schema.py:
```diff
@@ -330,13 +330,19 @@ def in_range(minimum: Optional[N], maximum: Optional[N], type_: Optional[TYPE] =
     None: {'type': 'null'}
 }
 
+_primitive_strings: Mapping[str, JSON] = {
```
bad name

amarjandu replied:
moved this reverse mapping into function where its used

hannes-ucsc commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,198 @@
+                                                  f" for '{facet}' must be one of '{valid_relations}'.")
+                        if not isinstance(value, schema.type_lookup(relation_type)):
+                            raise BadRequestError(
+                                msg=f"The value of the '{relation}' relation in the '{request_parameter['name']}'"
```
string interpolation
quotes

hannes-ucsc replied:
```suggestion
                                msg=f'The value of the {relation!r} relation in the {name!r}'
```

hannes-ucsc commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,198 @@
+                except KeyError:
+                    raise BadRequestError(f"Unknown facet '{facet}'")
+
+        if getattr(self, "content") is not None:
```
quotes
inner function just for if is bad
what does 'content' have to do with filters?

hannes-ucsc commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,198 @@
+                        raise BadRequestError(f"Invalid input for '{request_parameter}',"
+                                              f" must be within [{min_},{max_}]")
+
+        def _filter_validate(request_parameter: Mapping[str, str]):
```
should be generic, using jsonschema

hannes-ucsc reviewed (changes requested):
We should talk. My terse comments are meant to be reminders for myself to guide that convo.

jessebrennan commented on src/azul/chalice.py:
```diff
@@ -145,3 +159,122 @@ def _log_response(self, response):
+        parameters.extend(self.specs['paths'][resource_path].get(http_method, {}).get('parameters', []))
+        return parameters
+
+    def route(self,
```
This method is mostly just duplicated from the super class. Try and see if you can find a way to reduce that with a call to `super().route()`. It may involve refactoring the parent class's `__init__`.

amarjandu replied:
within the previous refactor i was trying to do something like;

```
Index: src/azul/chalice.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/azul/chalice.py	(revision a1bd2ef5517f26f5dd7c10080d0b24094fe89c3d)
+++ src/azul/chalice.py	(date 1597947274672)
@@ -2,6 +2,7 @@
 import json
 import logging
 from typing import (
+    Callable,
     Dict,
     Iterable,
     List,
@@ -59,6 +60,7 @@
               enabled: bool = True,
               path_spec: Optional[JSON] = None,
               method_spec: Optional[JSON] = None,
+              wrapped_function: Optional[Callable] = False,
               **kwargs):
         """
         Decorates a view handler function in a Chalice application.
@@ -90,15 +92,23 @@
 
             def decorator(view_func):
                 self._register_spec(path, path_spec, method_spec, methods)
+
                 # Stash the URL path a view function is bound to as an attribute of
-                # the function itself.
-                view_func.path = path
-                return chalice_decorator(view_func)
+                # the function itself
+                @functools.wraps(view_func)
+                def f(*args, **kwargs):
+                    if wrapped_function:
+                        wrapped_function()
+                    return view_func(*args, **kwargs)
+
+                f.path = path
+                return chalice_decorator(f)
 
             return decorator
         else:
             return lambda view_func: view_func
 
+
     def test_route(self, *args, **kwargs):
         """
         A route that's only enabled during unit tests.
@@ -231,50 +241,10 @@
               method_spec: Optional[JSON] = None,
               validate: bool = True,
               **kwargs):
-        """
-        Decorates a view handler function in a Chalice application.
-
-        See https://chalice.readthedocs.io/en/latest/api.html#Chalice.route.
-
-        :param path: See https://chalice.readthedocs.io/en/latest/api.html#Chalice.route
-
-        :param path_spec: Corresponds to an OpenAPI Paths Object. See
-                          https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md#pathsObject
-                          If multiple `@app.route` invocations refer to the same
-                          path (but with different HTTP methods), only specify
-                          this argument for one of them, otherwise an
-                          AssertionError will be raised.
-
-        :param method_spec: Corresponds to an OpenAPI Operation Object. See
-                            https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md#operationObject
-                            This should be specified for every `@app.route`
-                            invocation.
-
-
-        :param enabled: If False, do not route any requests to the decorated
-                        view function. The application will behave as if the
-                        view function wasn't decorated.
-
-        :param validate: If False, do not validate requests submitted to the endpoint.
-        """
-        if enabled:
-            methods = kwargs.get('methods', ())
-            chalice_decorator = super().route(path, **kwargs)
-
-            def decorator(view_func):
-                self._register_spec(path, path_spec, method_spec, methods)
-
-                # Stash the URL path a view function is bound to as an attribute of
-                # the function itself
-                @functools.wraps(view_func)
-                def f(*args, **kwargs):
-                    if validate:
-                        self.validate_params()
-                    return view_func(*args, **kwargs)
-
-                f.path = path
-                return chalice_decorator(f)
-
-            return decorator
-        else:
-            return lambda view_func: view_func
+        kwargs_to_pass = dict(path=path,
+                              enabled=enabled,
+                              path_spec=path_spec,
+                              method_spec=method_spec)
+        if validate:
+            kwargs_to_pass['wrapped_function'] = self.validate_params
+        super().route(**kwargs_to_pass)
\ No newline at end of file

```

But kept running into:
```
Failure
Traceback (most recent call last):
  File "/Users/amar/.pyenv/versions/3.8.2/lib/python3.8/unittest/suite.py", line 163, in _handleClassSetUp
    setUpClass()
  File "/Users/amar/dev/hca/azul/test/indexer/__init__.py", line 61, in setUpClass
    super().setUpClass()
  File "/Users/amar/dev/hca/azul/test/es_test_case.py", line 38, in setUpClass
    super().setUpClass()
  File "/Users/amar/dev/hca/azul/test/app_test_case.py", line 88, in setUpClass
    cls.app_module = load_app_module(cls.lambda_name(), unit_test=True)
  File "/Users/amar/dev/hca/azul/src/azul/modules.py", line 45, in load_app_module
    return load_module(path, f'lambdas.{lambda_name}.app', module_attributes)
  File "/Users/amar/dev/hca/azul/src/azul/modules.py", line 36, in load_module
    spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/Users/amar/dev/hca/azul/lambdas/service/app.py", line 375, in <module>
    def swagger_ui():
TypeError: 'NoneType' object is not callable
```
ill take another shot at this.

jessebrennan replied:
Here's a version: 
```
Index: src/azul/chalice.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/azul/chalice.py	(revision a1bd2ef5517f26f5dd7c10080d0b24094fe89c3d)
+++ src/azul/chalice.py	(date 1598312031206)
@@ -2,10 +2,12 @@
 import json
 import logging
 from typing import (
+    Callable,
     Dict,
     Iterable,
     List,
     Mapping,
+    MutableSequence,
     Optional,
 )
 
@@ -52,6 +54,7 @@
             self.specs['paths'] = {}
         else:
             self.specs: Optional[MutableJSON] = None
+        self.pre_call_hooks: MutableSequence[Callable[[], None]] = []
         super().__init__(app_name, debug=config.debug > 0, configure_logs=False)
 
     def route(self,
@@ -90,6 +93,8 @@
 
             def decorator(view_func):
                 self._register_spec(path, path_spec, method_spec, methods)
+                for hook in self.pre_call_hooks:
+                    hook()
                 # Stash the URL path a view function is bound to as an attribute of
                 # the function itself.
                 view_func.path = path
@@ -224,57 +229,10 @@
         parameters.extend(self.specs['paths'][resource_path].get(http_method, {}).get('parameters', []))
         return parameters
 
-    def route(self,
-              path: str,
-              enabled: bool = True,
-              path_spec: Optional[JSON] = None,
-              method_spec: Optional[JSON] = None,
-              validate: bool = True,
-              **kwargs):
+    def route(self, validate: bool = True, **kwargs):
         """
-        Decorates a view handler function in a Chalice application.
-
-        See https://chalice.readthedocs.io/en/latest/api.html#Chalice.route.
-
-        :param path: See https://chalice.readthedocs.io/en/latest/api.html#Chalice.route
-
-        :param path_spec: Corresponds to an OpenAPI Paths Object. See
-                          https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md#pathsObject
-                          If multiple `@app.route` invocations refer to the same
-                          path (but with different HTTP methods), only specify
-                          this argument for one of them, otherwise an
-                          AssertionError will be raised.
-
-        :param method_spec: Corresponds to an OpenAPI Operation Object. See
-                            https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md#operationObject
-                            This should be specified for every `@app.route`
-                            invocation.
-
-
-        :param enabled: If False, do not route any requests to the decorated
-                        view function. The application will behave as if the
-                        view function wasn't decorated.
-
         :param validate: If False, do not validate requests submitted to the endpoint.
         """
-        if enabled:
-            methods = kwargs.get('methods', ())
-            chalice_decorator = super().route(path, **kwargs)
-
-            def decorator(view_func):
-                self._register_spec(path, path_spec, method_spec, methods)
-
-                # Stash the URL path a view function is bound to as an attribute of
-                # the function itself
-                @functools.wraps(view_func)
-                def f(*args, **kwargs):
-                    if validate:
-                        self.validate_params()
-                    return view_func(*args, **kwargs)
-
-                f.path = path
-                return chalice_decorator(f)
-
-            return decorator
-        else:
-            return lambda view_func: view_func
+        if validate:
+            self.pre_call_hooks.append(self.validate_params)
+        super().route(**kwargs)
```

To take this idea to it's extreme, you could make `self._register_spec` be a `pre_call_hook` too. Then the signature for hooks would have to be changed to accept everything that's passed in to the route method. While this version would be more powerful and extendable, it might also be overkill.

amarjandu replied:
This seems to be in-lining callables at the same time the registration of the routes occur. We need to be performing the validation at a per-request basis, so it would have to be when the method that the route calls is invoked.

I'll take a look at this at a later time and focus on merging more code into the PathSpec class.

jessebrennan reviewed (changes requested):
I am envisioning a refactoring which would be difficult to describe as individual comments, so I'll try to put it concisely here. Currently, the validators are created for each request as they come in. This is inefficient when a lambda handles the same request multiple times.

To fix this, I propose changing the way the schema is stored in the ChaliceApp when the app is initialized. Consider first making a `Schema` class that is stored in the app in place of the raw schema. The `Schema` would still contain the raw schema, but also contain the functionality of your `SchemaValidation` class, i.e. the `Schema` class should have a method `validate()`. Then the `SchemaValidation` object could be removed.

Combining these classes will 1) cache everything that's needed to validate a request in the App object and 2) make validating something that the _schema_ does. The _schema_ will validate the request.

Some notes about the commits: 

* It's helpful when doing multi-part commits to have a body for the commit message that explains what's in each part. (@hannes-ucsc, should this be added to the contributing guide?)
  
  For example `[1/3] Validate requests using swagger spec #(1465)` could have a commit body that says, `Remove old validation code`.

* This is probably more of a hassle than it's worth to fix, but the `Swagger` should be capitalized in the commits

There's kind of a lot here, and it may not all be clear. Let me know on Slack if you have more questions and/or want to talk about it.

amarjandu commented on src/azul/chalice.py:
```diff
@@ -145,3 +159,122 @@ def _log_response(self, response):
+                parameter_schema = parameter['schema']
+            elif 'content' in parameter:
+                for k, v in parameter['content'].items():
+                    parameter_schema = v['schema']
```
per the spec for https://swagger.io/specification/#parameter-schema, 
`A parameter MUST contain either a schema property, or a content property, but not both.`

I replaced with an `else` and placed an assertion for 'content'

amarjandu commented on src/azul/chalice.py:
```diff
@@ -145,3 +159,122 @@ def _log_response(self, response):
+
+        query_params = self.current_request.query_params if self.current_request.query_params else {}
+        uri_params = self.current_request.uri_params if self.current_request.uri_params else {}
+        all_provided_params = {**query_params, **uri_params}
```
as of now im opting not to inline the variable. Doing so allows me to do a set intersection and assignment (&=) for the parameters to inspect.

amarjandu replied:
Ill take another look at this before i request review.

jessebrennan commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,109 @@
+            try:
+                filter_json = json.loads(param_value)
+            except Exception:
+                raise BadRequestError(f'The {param_name!r} parameter is not valid JSON')
```
This error doesn't have a period at the end of it while the rest all do. Skimming the codebase, it doesn't seem like we have a strict convention on this, but I see a slight preference of no period.

Regardless, I think within a single file, errors should be consistent, so either just add one here, or remove the others.

amarjandu replied:
Looking around at other APIs some appear to prefer punctuation and others don't. 
No Punctuation - https://developers.coinbase.com/api/v2#errors
Has Punctuation -  https://developers.google.com/drive/api/v3/handle-errors , https://stripe.com/docs/declines#issuer-declines 

I think we should consistently perform one, just browsing through the [openapi](https://service.dev.singlecell.gi.ucsc.edu/openapi)  endpoint it looks like we have mixed usage. Because these messages are intended to be for humans to read I would be inclined to include the punctuation, (and establish it in the contributing guidelines). There might be some refactoring to do in the API descriptions if we perform this.

@hannes-ucsc do you have input/preference?

amarjandu replied:
@hannes-ucsc "no periods if it is a single sentence or fragment. If there are  multiple sentences, then there needs to be a period at the end."

amarjandu replied:
@hannes-ucsc  "when making small changes you should follow precedence for what is already there, if you are making large changes you should update the module to the new guidelines."

hannes-ucsc replied:
This was already covered. An upcoming PR will add a section title to the guidelines that already cover this. The title will be *Consistency and precedent*

amarjandu commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,109 @@
+                    if len(e.absolute_schema_path) == 1 and 'additionalProperties' in e.absolute_schema_path:
+                        raise BadRequestError(f'Unknown facet {facet!r}')
+                    elif 'oneOf' in e.absolute_schema_path and 'additionalProperties' in e.absolute_schema_path:
+                        valid_relations = [list(x['properties'].keys())[0] for x in e.parent.schema['oneOf']]
```
```json
{
   "oneOf":[
      {
         "type":"object",
         "properties":{
            "is":{
               "type":"array",
               "items":{
                  
               }
            }
         },
         "additionalProperties":False,
         "required":[
            "is"
         ]
      },
      {
         "type":"object",
         "properties":{
            "contains":{
               "type":"array",
               "items":{
                  "type":"array",
                  "items":{
                     
                  },
                  "minItems":2,
                  "maxItems":2
               },
               "minItems":1,
               "maxItems":2
            }
         },
         "additionalProperties":False
      },
      {
         "type":"object",
         "properties":{
            "within":{
               "type":"array",
               "items":{
                  "type":"array",
                  "items":{
                     
                  },
                  "minItems":2,
                  "maxItems":2
               },
               "minItems":1,
               "maxItems":2
            }
         },
         "additionalProperties":False
      },
      {
         "type":"object",
         "properties":{
            "intersects":{
               "type":"array",
               "items":{
                  "type":"array",
                  "items":{
                     
                  },
                  "minItems":2,
                  "maxItems":2
               },
               "minItems":1,
               "maxItems":2
            }
         },
         "additionalProperties":False
      }
   ]
}
``` 
It's kinda nasty to get the valid relation values out, I'm inclined to leave it as is, It should also handle API changes when we add fix, https://github.com/DataBiosphere/azul/issues/2167 .

jessebrennan replied:
Also rename variable from `x` to something more representative.

jessebrennan reviewed (changes requested):
Might want to check with @hannes-ucsc about this first, but I think it would be helpful to consolidate the history at this point.

amarjandu commented on src/azul/openapi/validation.py:
```diff
@@ -35,6 +39,13 @@ def __init__(self, **kwargs):
+        if getattr(self, 'schema'):
+            parameter_schema = self.schema
+        else:
+            assert getattr(self, 'content')
```
I ended up refactoring the __init__ class due to how the attributes were being set. Any instance of the class would have the attribute `schema` and `content` but sometimes the values were not set.

This was set to an assertion that the attribute is not None.

amarjandu commented on src/azul/chalice.py:
```diff
@@ -277,8 +283,10 @@ def fmt_error(err_description, parameters):
 
-        query_params = current_request.query_params if current_request.query_params else {}
-        uri_params = current_request.uri_params if current_request.uri_params else {}
+        query_params = current_request.query_params if (current_request.query_params
```
TODO format these 2 lines.

jessebrennan commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,141 @@
+                    try:
+                        validate_uuid(instance)
+                    except (InvalidUUIDError, InvalidUUIDVersionError):
+                        yield jsonschema.exceptions.ValidationError('%r is an invalid UUID' % instance)
```
I'm super confused by this yield statement. Normally we just raise exception instances. I didn't see anything in the jsonschema docs about this (but I didn't look too hard). What's going on here?

amarjandu replied:
i recall reading in the jsonschema documentation that a yield statement should be used. I cant locate it at the moment, and upon inspecting https://python-jsonschema.readthedocs.io/en/latest/errors/?#jsonschema.exceptions.ValidationError a return or raise can be used.

amarjandu replied:
https://github.com/Julian/jsonschema/blob/master/jsonschema/_validators.py

amarjandu replied:
not using a yield statement fails to set information about the exception being raised, this causes issues with figuring out what caused the issue.

jessebrennan commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,141 @@
+                param_value = int(param_value)
+            elif self.schema.get('type') == 'array':
+                param_value = param_value.split(',')
+            else:
+                pass
```
If comment above holds, remove the final `else: pass` here too.

jessebrennan commented on test/service/test_request_validation.py:
```diff
@@ -249,19 +270,111 @@ def test(url, message, params):
+                         message="The 'filters' parameter is not valid JSON")
+        if "/integrations" in self.app_module.app.specs['paths'].keys():
+            self.fail(msg="Remove conditional check to allow parameters for '/integrations' to be tested")
+            # TODO place this subtest back when the `/integrations` endpoint is spec'd out.
```
Probably should connect this TODO with a ticket so that whoever specs `/integrations` knows to fix this too.

jessebrennan commented on test/service/test_request_validation.py:
```diff
@@ -249,19 +270,111 @@ def test(url, message, params):
-                 params={},
-                 message='Missing required query parameters `entity_type`, `integration_type`')
+                         message="The 'filters' parameter is not valid JSON")
+        if "/integrations" in self.app_module.app.specs['paths'].keys():
```
Nitpick:
```suggestion
        if '/integrations' in self.app_module.app.specs['paths'].keys():
```

amarjandu replied:
not a nit, its a requirement from contribution guidelines ha.

jessebrennan commented on test/service/test_request_validation.py:
```diff
@@ -249,19 +270,111 @@ def test(url, message, params):
+        facet = 'organPart'
+        for relation in ['contains', 'within', 'intersects']:
+            path_specs = self.app_module.app.route_specs[path].extract_parameters('get')
+            filter_spec = one(x for x in path_specs if x['name'] == 'filters')
```
Chose something more meaningful than `x`.

jessebrennan commented on test/service/test_request_validation.py:
```diff
@@ -249,19 +270,111 @@ def test(url, message, params):
+                'properties',
+                facet,
+                'oneOf']
+            relation_spec = one(x for x in reduce(getitem, spec_location, filter_spec)
```
Chose something more meaningful than `x`.

jessebrennan commented on test/service/test_request_validation.py:
```diff
@@ -249,19 +270,111 @@ def test(url, message, params):
+                'schema',
+                'properties',
+                facet,
+                'oneOf']
```
```suggestion
                'oneOf'
            ]
```

jessebrennan commented on src/azul/chalice.py:
```diff
@@ -145,3 +159,122 @@ def _log_response(self, response):
+        uri_params = self.current_request.uri_params if self.current_request.uri_params else {}
+        all_provided_params = {**query_params, **uri_params}
+
+        provided_params = set(all_provided_params.keys())
```
It looks like you never addressed this, but still gave the thumbs up :thinking: 

> The set() call is unnecessary here, since a dictionary view object implements most (all?) set methods. Also I would probably just inline this variable.

https://github.com/DataBiosphere/azul/pull/1849/files#r471770897

jessebrennan reviewed (changes requested):
Feel free to consolidate the history again before adding these changes.

I decided to check out the coverage reports, but it doesn't look like anything changed, which is surprising to me. It didn't investigate much further, but give it a look if you have a chance.

amarjandu commented on src/azul/openapi/schema.py:
```diff
@@ -192,12 +193,31 @@ def enum(*items: PrimitiveJSON, type_: TYPE = None) -> JSON:
+    else:
+        require(len(items_) == 0)
+        #TODO this isinstance check needs to be looser. allow mappings. 
+        #require(isinstance(items, list))
```
Very strange, in my working copy this space is present.

amarjandu commented on lambdas/service/app.py:
```diff
@@ -896,7 +896,7 @@ def repository_search(entity_type: str, item_id: str):
                         schema.object_type(
                             properties={
-                                op: schema.array({}, minItems=2, maxItems=2)
+                                op: schema.array({},items=schema.array({}, minItems=2, maxItems=2), minItems=1, maxItems=2)
```
Super strange, this also was formatted correctly..... interesting that this question was marked with 'OUTDATED' up top, maybe your browser needed to be refreshed when doing the review?

jessebrennan replied:
For the record, this was because I was reviewing "per-commit", and the changes were updated in a later commit.

amarjandu commented on src/azul/openapi/schema.py:
```diff
@@ -211,8 +211,8 @@ def enum(*items_: PrimitiveJSON, items: List[str] = None, type_: TYPE = None) ->
-        #TODO this isinstance check needs to be looser. allow mappings. 
-        #require(isinstance(items, list))
+        # TODO this isinstance check needs to be looser. allow mappings.
+        # require(isinstance(items, list))
```
Duplicate

amarjandu commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,141 @@
+            self.schema = self.content['application/json']['schema']
+            try:
+                filter_json = json.loads(param_value)
+            except Exception:
```
This was brought over from https://github.com/DataBiosphere/azul/blob/4e4e231d5f7205217321003cf9e9898a7b825243/lambdas/service/app.py#L693

However I'm open to narrowing down the exception.

jessebrennan replied:
It seems like the earlier mistake was overlooked. Don't you think we should correct it?

amarjandu commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,141 @@
+            except Exception:
+                raise BadRequestError(f'The {param_name!r} parameter is not valid JSON')
+            if type(filter_json) is not dict:
+                raise BadRequestError(f'The {param_name} parameter must be a dictionary')
```
Same as above, but im open for this change.

amarjandu commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,141 @@
+                for facet, filter_ in filter_json.items():
+                    try:
+                        relation, value = one(filter_.items())
+                    except Exception:
```
This was also migrated over from the old codebase. I'm open to the specification. going to ping @hannes-ucsc to ensure this was not intentional.

amarjandu commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,147 @@
+            except (TypeError, ValueError):
+                raise BadRequestError(msg=f'Invalid value for parameter {param_name!r}')
+            assert self.schema is not None
+            if self.schema.get('format') == 'uuid':
```
check for type == str, refactor type into variable.

amarjandu commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,147 @@
+                param_value = int(param_value)
+            elif self.schema.get('type') == 'array':
+                param_value = param_value.split(',')
+            else:
```
remove this else pass.

amarjandu commented on lambdas/service/app.py:
```diff
@@ -2207,16 +2213,20 @@ def cart_export_send_to_collection_api(event, context):
+               'description': format_description('''
         This endpoint returns object metadata, and a list of access methods that can
         be used to fetch object bytes.
     ''') + drs_spec_description,
```
per our conversation Friday, modifications for this have been pushed up

jessebrennan commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,147 @@
+                    try:
+                        jsonschema.validate({facet: filter_}, self.schema)
+                    except (
+                        jsonschema.exceptions.SchemaError,
```
This is a major bug. `SchemaError` is only raised when "an invalid schema itself is encountered". If any of our schemas are invalid you are reporting them as 4XX errors.

If you are seeing `SchemaError`s then there's either a bug in the schemas or your conversion code. Either way, I don't think this is an error we should catch.

You *should* add a test that checks each schema and asserts that a `SchemaError` is *not* raised.

Check elsewhere for this same bug.

amarjandu replied:
Added check_schema to ServiceSpecValidation test

amarjandu replied:
the schema is validated during the IT, moved it to be a unit-test for easier testing.

jessebrennan commented on src/azul/chalice.py:
```diff
@@ -145,3 +161,160 @@ def _log_response(self, response):
+        if path not in self.route_specs:
+            assert path not in self.specs['paths'], 'Only specify path_spec once per route path'
+            route_spec = RouteSpec(path, path_spec)
+            self.route_specs[path] = route_spec
```
PyCharm shows a type warning here. Let me know if you need help resolving this. Make sure that there are no more type errors in this area after fixing this one.

amarjandu replied:
Changed self.route_specs to `MutableMapping` pycharm code inspection does not find any new warnings.

jessebrennan commented on src/azul/chalice.py:
```diff
@@ -145,3 +161,160 @@ def _log_response(self, response):
+        """
+        Returns a list of OpenAPI `parameters` that are applicable to a given API route
+        """
+        return [parameter for parameter in chain(self.path_spec.get('parameters', []),
+                                                 self.get_method_spec(http_method).get('parameters', []))]
```
```suggestion
        return [*self.path_spec.get('parameters', []), *self.get_method_spec(http_method).get('parameters', []))]
```

It looks like you can remove the `chain` import now too.

jessebrennan reviewed (changes requested):
Some of my comments here are out of date because I wrote them before you pushed some changes :shrug:. Feel free to just thumb those up and I'll mark them resolved later.

amarjandu commented on test/service/test_request_validation.py:
```diff
@@ -121,7 +120,10 @@ def test_bad_sort_facet_and_filter_facet_of_sample(self):
         response = requests.get(url, params=params)
         self.assertEqual(400, response.status_code, response.json())
-        self.assertIn(response.json(), [self.facet_message, self.facet_enumeration_message()])
+        try:
+            self.assertEqual(self.facet_message, response.json())
+        except AssertionError:
+            self.assertIn(self.facet_enum_message, response.json()['Message'])
```
I had to use try: except because if there was a failing in one of the tests, the OR statement does not get evaluated, the test failure is raised.

```
Traceback (most recent call last):
  File "/Applications/PyCharm.app/Contents/plugins/python/helpers/pycharm/teamcity/diff_tools.py", line 32, in _patched_equals
    old(self, first, second, msg)
  File "/Users/amar/.pyenv/versions/3.8.2/lib/python3.8/unittest/case.py", line 912, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/Users/amar/.pyenv/versions/3.8.2/lib/python3.8/unittest/case.py", line 1211, in assertDictEqual
    self.fail(self._formatMessage(msg, standardMsg))
  File "/Users/amar/.pyenv/versions/3.8.2/lib/python3.8/unittest/case.py", line 753, in fail
    raise self.failureException(msg)
AssertionError: {'Cod[47 chars]ror: Unknown facet 'bad-facet'"} != {'Cod[47 chars]ror: Invalid input for 'sort'. Must be one of [794 chars]']."}
Diff is 1201 characters long. Set self.maxDiff to None to see it.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/amar/.pyenv/versions/3.8.2/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/Users/amar/.pyenv/versions/3.8.2/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/Users/amar/.pyenv/versions/3.8.2/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "/Users/amar/dev/hca/azul/test/service/test_request_validation.py", line 124, in test_bad_sort_facet_and_filter_facet_of_sample
    (self.assertEqual(self.facet_message, response.json()) or
```

jessebrennan replied:
Your reasoning makes sense. I'm wondering now why the assert error is being raised in the first place. Assertions being raised indicate a problem with the code, so catching one in a test is a bad smell. But maybe I'm misunderstanding what's going on here.

amarjandu commented on lambdas/service/app.py:
```diff
@@ -2244,7 +2008,8 @@ def assert_jwt_ttl(expected_ttl):
 
-@app.test_route('/resources/carts/{cart_id}/export', methods=['GET', 'POST'], cors=True, authorizer=jwt_auth)
+@app.test_route('/resources/carts/{cart_id}/export', methods=['GET', 'POST'], cors=True, authorizer=jwt_auth,
+                validate=False)
```
this and additional locations were changes during your review.

amarjandu commented on lambdas/service/app.py:
```diff
@@ -2285,7 +2050,8 @@ def export_cart_as_collection(cart_id: str):
 
-@app.test_route('/fetch/resources/carts/{cart_id}/export', methods=['GET', 'POST'], cors=True, authorizer=jwt_auth)
+@app.test_route('/fetch/resources/carts/{cart_id}/export', methods=['GET', 'POST'], cors=True, authorizer=jwt_auth,
+                validate=False)
```
this and additional locations were changes during your review.

amarjandu commented on lambdas/service/app.py:
```diff
@@ -2094,7 +1857,8 @@ def delete_cart_item(cart_id, item_id):
 
-@app.test_route('/resources/carts/{cart_id}/items/batch', methods=['POST'], cors=True, authorizer=jwt_auth)
+@app.test_route('/resources/carts/{cart_id}/items/batch', methods=['POST'], cors=True, authorizer=jwt_auth,
+                validate=False)
```
this and additional locations were changes during your review.

amarjandu commented on lambdas/service/app.py:
```diff
@@ -598,233 +607,9 @@ def version():
-
-@app.route('/integrations', methods=['GET'], cors=True)
+@app.route('/integrations', methods=['GET'], cors=True, validate=False)
 def get_integrations():
```
Added reference to $1984

amarjandu commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,147 @@
+                 schema: Mapping = None,
+                 **kwargs):
+        """
+        :param kwargs: Used to extract the `in` parameter.
```
unexpected kwargs would cause failure in the integration tests. due to a mismatch in the Open-API specification. 

https://github.com/DataBiosphere/azul/blob/32b462d2f32d3de3f17b3a3a2128515bb8a3a6d6/test/integration_test.py#L568

I do think we can perform the validation at the unit-test level rather than waiting for the integration-test to be run // infra to be deployed.. Because the spec is the same as the runtime, we don't have to inspect for unexpected kwargs.

amarjandu replied:
added openapi spec validation test to ServiceSpecValidation, removed from IT

amarjandu commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,147 @@
+                    except (ValueError, AttributeError):
+                        raise BadRequestError(
+                            f"The 'filters' parameter entry for {filter_!r} must be a single-item"
+                            f" dictionary")
```
> Also the second line should not be a format string.

I'm confused about this statement. can you explain?

jessebrennan commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,144 @@
+                        relation, value = one(filter_.items())
+                    except ValueError:
+                        raise BadRequestError(f"The 'filters' parameter entry for {facet!r} may only "
+                                              f"specify a single relation")
```
I removed the f in `f"..."` because it's not necessary on this line. Same thing with line 83.

amarjandu replied:
although its not necessary, im not a fan of how it looks when the 'f' is removed on some lines and not others.
I brought this up with Hannes his response was,
 @hannes-ucsc "f on every line"

jessebrennan commented on test/service/test_schema_validation.py:
```diff
@@ -0,0 +1,24 @@
+from furl import (
+    furl,
+)
+from jsonschema import (
+    Draft4Validator,
+)
+from openapi_spec_validator import (
+    validate_spec,
+)
+import requests
+
+from service import (
+    WebServiceTestCase,
+)
+
+
+class ServiceSpecValidation(WebServiceTestCase):
+
+    def test_validate_spec(self):
+        response = requests.get(url=furl(url=self.base_url, path='openapi').url)
+        response.raise_for_status()
+        spec = response.json()
+        validate_spec(spec)
+        Draft4Validator.check_schema(spec)
```
Consider moving the contents of this file to `test/test_openapi.py`

jessebrennan commented on test/service/test_schema_validation.py:
```diff
@@ -0,0 +1,24 @@
+    furl,
+)
+from jsonschema import (
+    Draft4Validator,
```
Why did you switch from Draft 7 to Draft 4?

amarjandu replied:
Open API 3.0.3 is based on draft‑wright‑json‑schema‑00 , which has a metaschema identifier of `draft-04`
https://json-schema.org/specification-links.html
https://swagger.io/specification/#data-types

I had used the incorrect one the first time around.

amarjandu commented:
> Your reasoning makes sense. I'm wondering now why the assert error is being raised in the first place. Assertions being raised indicate a problem with the code, so catching one in a test is a bad smell. But maybe I'm misunderstanding what's going on here.

@jessebrennan There's no comment box above so ill respond here. The request that test is sending has 2 bad parameters begin passed into the system, so we expect one of the two messages to be returned from the API. We have to catch the assertion error if the error message returned from does not match, then inspect if the message matches the other API response that can happen.

jessebrennan reviewed (changes requested):
I pushed a commit with minor formatting changes (mostly from the diff I sent you before which was incorrect :facepalm: . Feel free to just merge it back into your stuff. 

![amar-review](https://user-images.githubusercontent.com/10734740/94614850-a8128e80-025b-11eb-935d-6459d3211531.png)

Please address the first two warnings and the misspelling of parameters. The "shadowed name" warnings are fine to ignore I think.

It's probably a good idea to scan for more PyCharm errors like these in the files you've changed. I noticed these, but didn't look too thoroughly.

amarjandu commented:
https://travis-ci.org/github/DataBiosphere/azul/jobs/733063078 
Seems like the spec for the `/dss/` endpoints were removed from develop. Trying to resolve test errors. 
Removal of DSS endpoints tracked here.
https://github.com/databiosphere/azul/issues/2311

jessebrennan commented on src/azul/openapi/validation.py:
```diff
@@ -113,15 +115,16 @@ def validate(self, request_parameter: Mapping[str, str]):
             schema_type = self.schema['type']
             if schema_type == 'string':
                 if self.schema.get('format') == 'uuid':
+                    # noinspection PyUnusedLocal
```
I didn't intend to instruct you to add this `noinspection`. I don't think it's necessary since we don't ignore this warning in other places, but if you want to leave it in, go ahead.

amarjandu replied:
>Please address the first two warnings and the misspelling of parameters. The "shadowed name" warnings are fine to ignore I think.

I thought the `not used parameters` fell into the case of warnings to address.  It looks like there is other usage of this in the codebase, ill be keeping it in.

amarjandu commented on src/azul/openapi/validation.py:
```diff
@@ -47,8 +47,10 @@ def __init__(self,
+            # noinspection PyUnresolvedReferences
             for k, v in self.content.items():
                 parameter_schema = v['schema']
+        # noinspection PyUnboundLocalVariable
```
I'm debating if i should remove this check, as the openapi schema validation tests would throw errors if the API schema is defined in a way where this would raise a type error.

jessebrennan commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,149 @@
+        self.schema = schema
+        self.json_validators = dict(jsonschema.Draft4Validator.VALIDATORS)
+        self._in = kwargs['in']
+        if self.schema is not None and self.content is not None:
+            raise TypeError('OpenAPI parameter specification enforces that schema'
+                            ' and content are mutually exclusive')
+        if self.schema:
+            self.validator = type_lookup(self.schema['type'])
+        elif self.content:
+            self.validator = type_lookup(self.content['application/json']['schema']['type'])
+        else:
+            raise TypeError('Parameters must contain either a schema property or'
+                            ' content property')
```
A slightly more succinct way to write this would be:
```suggestion
        if self.schema is not None and self.content is None:
            self.validator = type_lookup(self.schema['type'])
        elif self.content is not None and self.schema is None:
            self.validator = type_lookup(self.content['application/json']['schema']['type'])
        else:
            raise TypeError('Parameters must contain either schema or content'
                            '(but not both)')
```

jessebrennan commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,149 @@
+        elif self.content:
+            self.validator = type_lookup(self.content['application/json']['schema']['type'])
+        else:
+            raise TypeError('Parameters must contain either a schema property or'
```
I might have used a `ValueError` for this, but actually I'm not sure which is more correct. So this isn't a suggestion, but I thought I'd mention it.

amarjandu replied:
after reading https://docs.python.org/3.8/library/exceptions.html?highlight=typeerror#TypeError
```
Passing arguments of the wrong type (e.g. passing a list when an int is expected) should result in a TypeError, but passing arguments with the wrong value (e.g. a number outside expected boundaries) should result in a ValueError.
```

We have no inspected if the values are in the expected boundaries, just that they are not set correctly (exclusively). So a TypeError seemed to fit.

jessebrennan commented on src/azul/chalice.py:
```diff
@@ -145,3 +158,163 @@ def _log_response(self, response):
+                raise BadRequestError(msg=fmt_error('Missing required',
+                                                    sorted(missing_params)))
+        for param_name in provided_params:
+            if param_name in validation_params:
```
My guess is that doing the set operation scrambled the order or the parameters? I just want to confirm that I understand what was going on in the history.

amarjandu replied:
Yeah that's correct, the set operation `&=` has messed with the ordering of the variables, I patched this on `develop` as well. https://github.com/DataBiosphere/azul/pull/2359

jessebrennan reviewed (commented):
This is just a review of the latest changes which look :+1:. I'm going to do a separate review of everything before I give approval.

jessebrennan commented on src/azul/chalice.py:
```diff
@@ -145,3 +158,163 @@ def _log_response(self, response):
+
+    @property
+    def methods(self):
+        return [method for method in self.specs.keys()]
```
One problem with generating a list here is that a new list will be allocated in memory each time this property is accessed. A simpler solution would just be
```suggestion
        return self.specs.keys()
```
which only returns a reference to the same object.

This new method not only saves memory space, but it's faster too. The property is used to check membership which is O(n) on a list, but O(1) on a set or dict (or dict view object in this case).

amarjandu replied:
That's a good catch, ill try to use this more often.

jessebrennan reviewed (changes requested):
Super close, just a couple things.

amarjandu commented on src/azul/openapi/schema.py:
```diff
@@ -192,12 +193,31 @@ def enum(*items: PrimitiveJSON, type_: TYPE = None) -> JSON:
+        items_ = list(items_)
+    else:
+        require(len(items_) == 0)
+        # TODO: This isinstance check needs to be looser. allow mappings.
+        # require(isinstance(items, list))
```
IIRC - this was something that @hannes-ucsc requested that I add in when changing the enumerations from sets to lists.

I can create an issue for this....

amarjandu commented on lambdas/service/app.py:
```diff
@@ -1661,6 +1661,10 @@ def fetch_dss_files(file_uuid: str) -> Response:
     }
 })
 def repository_files(file_uuid: str) -> Response:
+    return repository_files_worker(file_uuid)
+
+
+def repository_files_worker(file_uuid: str) -> Response:
```
This was introduced to get around https://github.com/DataBiosphere/azul/pull/1849#issuecomment-704402843

amarjandu commented on src/azul/openapi/schema.py:
```diff
@@ -231,6 +231,17 @@ def enum(*items_: PrimitiveJSON, items: List[str] = None, type_: TYPE = None) ->
+    return {
+        **make_type(str),
+        'format': 'uuid'
+
```
rm line here

jessebrennan reviewed (changes requested):
If there is an error from missing parameters and an error from validation, the user will only get an error for missing parameters.

Since you are making an effort to show all errors together, it makes sense that all of the errors, even if they are of different kinds (missing vs validation) should be shown together. 

Having said that, this error logging code is getting pretty verbose and complicated. It might be simpler just to leave it how it was before.

This review only addresses the latest commit. I'm going to review further in a separate "Github" review.

amarjandu commented on src/azul/chalice.py:
```diff
@@ -275,10 +277,11 @@ def _create_spec_validators(self, http_method: str) -> Mapping[str, SpecValidato
-            return f'{err_description} {pluralize("parameter", len(parameters))} {joined}'
+    def validate_params(self, current_request: Request) -> Union[Response, None]:
+        def bad_response(body: Mapping):
+            return Response(body=json.dumps(body),
```
removed from body.

amarjandu commented:
@jessebrennan I made the changes you've requested above in my local working copy, I'll wait for you to finish the review before pushing. 
In the case of merging the invalid, missing, and extra parameters, which `title` should be used within the response?

jessebrennan reviewed (changes requested):
This was the only other thing I found.

amarjandu commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,159 @@
+            self.validator = type_lookup(self.schema['type'])
+        elif self.content is not None and self.schema is None:
+            self.validator = type_lookup(self.content['application/json']['schema']['type'])
+        else:
```
i was not going to do it, but for you, i added it. :)

jessebrennan replied:
You don't have to do it if you don't agree, but I didn't see a comment or explanation either.

amarjandu replied:
I was reluctant to add the changes, but after doing it i think it looks a lot better. :)

jessebrennan commented:
>In the case of merging the invalid, missing, and extra parameters, which title should be used within the response?

You could consider reformatting your JSON structure to be something like:

```
{
    "errors": [
        {
            "title": "foo",
            "message": "bar",
            etc.
        }
    ]
}
```

jessebrennan commented on src/azul/openapi/validation.py:
```diff
@@ -45,119 +53,165 @@ def __init__(self,
-                                      schema=e.schema)
+                            value_and_unit.to_index(value)
+                        except RequirementError as e:
+                            raise ValidationError(message=e.message,
```
PyCharm shows a type error for `e.message`.

amarjandu commented on src/azul/chalice.py:
```diff
@@ -145,3 +159,189 @@ def _log_response(self, response):
+
+        if mandatory_params:
+            missing_params = mandatory_params - provided_params
+            if missing_params:
```
currently we don't have any query parameters are required parameters. The required parameters are path parameters, but API gateway returns an error before we get a response from the lambda. This is also emulated on `chalice local` so we are not able to run unit-tests that skip path parameters. I raised https://github.com/DataBiosphere/azul/issues/2465, and added a skipped test that references the issue.

amarjandu commented:
@jessebrennan I don't think removing the old usage for providing default values to missing parameters should be done yet. There are certain endpoints that are being deprecated and don't get the benefits of the new spec parameter validation. The codebase would appear to be fragmented in how query parameters are expected, eg. 

If there is a missing catalog parameter for the `/dss/` and `/drs/` endpoints, when the function `_repository_files` calls `app.repository_controller.download_file(catalog=app.catalog,...` the `/dss/` endpoint relies on the `app.catalog` property to fill in a default value, while the `/drs/` endpoint gets a default catalog from the new validation methods. 

There is some additional code-debt that would be created if we want to include this feature, so I understand if we can defer its addition to a later time. I do think that having the code-debt is not necessarily a bad thing in this case, as the feature is seems useful, and the codebase currently can handle the changes appropriately.

jessebrennan reviewed (changes requested):
I pushed a commit with my suggestions to make it easier to implement if you agree with them.

amarjandu commented on src/azul/chalice.py:
```diff
@@ -145,3 +160,202 @@ def _log_response(self, response):
+            all_provided_params.update({
+                name: spec_parameters[name].default for name in default_params
+            })
+            current_request.query_params = all_provided_params
```
The change was intentional to reduce time wasted on validating parameters that are set as defaults, i think we can add a test that checks this instead.

amarjandu commented:
@jessebrennan 
I pushed 1 commit that cleans up the object that is passed to `spec_parameter.validate()`. 
In the second commit I fixed an error within the suggestion, and added some additional testing, I moved the setting of the default parameters to after all the validation is complete to speed up runtime.

jessebrennan commented on src/azul/chalice.py:
```diff
@@ -145,3 +160,187 @@ def _log_response(self, response):
+                            status_code=400,
+                            headers={'Content-Type': 'application/problem+json'})
+
+        for params, kind in ((current_request.query_params, 'query'),
```
Don't you need to add the defaults before you check the missing params?

jessebrennan reviewed (changes requested):
Your changes didn't solve the problem. Make sure you test it before sending it back for review.

jessebrennan commented on test/service/test_request_validation.py:
```diff
@@ -236,45 +268,203 @@ def test_file_order(self):
+                       'responses': {'200': {'description': 'OK'}}, 'tags': ['Index']}
+
+        for required, catalog in product((True, False), (self.catalog, None)):
+            param_spec = {'in': 'query', 'schema': {
+                'type': 'string',
+                'pattern': '^([a-z0-9]{1,64})$',
+                'enum': ['test']
+            }, 'required': required, 'name': f'required-{required}-default-{catalog}'}
+            if catalog is not None:
+                param_spec['schema']['default'] = catalog
```
I think Hannes generally prefers the declarative style for building dict literals like this. Also notice that I changed the dict literal formatting.
```suggestion
            param_spec = {
                'in': 'query',
                'schema': {
                    'type': 'string',
                    'pattern': '^([a-z0-9]{1,64})$',
                    'enum': ['test']
                },
                'required': required, 
                'name': f'required-{required}-default-{catalog}'
                **({} if catalog is None else {'default': catalog})
            }
```

amarjandu replied:
This had to be modified slightly, but 👍🏽

jessebrennan commented on test/service/test_request_validation.py:
```diff
@@ -236,45 +268,203 @@ def test_file_order(self):
+
+    def test_default_for_missing_params(self):
+        path = '/test/mock/endpoints'
+        method_spec = {'parameters': [],
+                       'responses': {'200': {'description': 'OK'}}, 'tags': ['Index']}
```
With the changes below, see if you can now make this into a list comprehension. Also, see formatting guide for JSON literals.

jessebrennan commented on test/service/test_request_validation.py:
```diff
@@ -236,45 +268,203 @@ def test_file_order(self):
+        url = furl(url=self.base_url, path=path).url
+        response = requests.get(url=url)
+        self.assertEqual(400, response.status_code, response.json())
+        self.assertEqual('required-True-default-None',
+                         response.json()['missing_parameters'][0]['name'])
```
Using one (from itertools) would remove the need for the assert after this one.

```suggestion
        self.assertEqual('required-True-default-None',
                         one(response.json()['missing_parameters'])['name'])
```

jessebrennan reviewed (changes requested):
I like the test! I found a few cosmetic things.

jessebrennan commented on test/service/test_request_validation.py:
```diff
@@ -458,7 +460,7 @@ def test_method():
         self.assertEqual('required-True-default-None',
-                         response.json()['missing_parameters'][0]['name'])
+                         one(response.json()['missing_parameters'])['name'])
         self.assertEqual(1, len(response.json()['missing_parameters']))
```
This line is now redundant and thus should be deleted. I tried to make this clear in my previous review.

amarjandu replied:
doh, removed it from other locations and forgot about this one. :(

jessebrennan commented on test/service/test_request_validation.py:
```diff
@@ -433,18 +427,26 @@ def test_missing_uuid(self):
+                    },
+                    'required': required,
+                    'name': f'required-{required}-default-{default}'
+                } for required, default in product((True, False), (self.catalog, None))
```
This is purely suggestion, but this method is also slightly shorter.
```suggestion
                } for required in (True, False) for default in (self.catalog, None)
```

jessebrennan reviewed (changes requested):
Almost there.

Can you clean up the history before adding these changes?

jessebrennan reviewed (approved):
Nice work! Please clean up the history.

hannes-ucsc commented on src/azul/openapi/schema.py:
```diff
@@ -338,7 +338,7 @@ def in_range(minimum: Optional[N], maximum: Optional[N], type_: Optional[TYPE] =
 
 
-def object_type(properties: JSON, **kwargs) -> JSON:
+def object_type(properties: JSON, additional_properties=False, **kwargs) -> JSON:
```
This goes against the philosophy behind the `…_type` functions. They don't treat any schema properties special by elevating them to keyword arguments. Also note the call to this in line 93 that still passes this in kwargs rather than the special keyword argument. Also, while you're touching this: `properties` should be a positional only argument.

So

```suggestion
def object_type(properties: JSON, /, **kwargs) -> JSON:
```

hannes-ucsc commented on lambdas/service/app.py:
```diff
@@ -953,9 +953,11 @@ def repository_search(entity_type: str, item_id: Optional[str]) -> JSON:
-                            op: schema.array({}, minItems=2, maxItems=2)
-                        })
+                        schema.object_type(
+                            properties={
```
and that will prevent you from doing this.

hannes-ucsc reviewed (changes requested):
Reviewing commit by commit. This review is for the first one.

Please motivate, in parking lot, why you'd like this change.

theathorn commented:
Amar to resolve conflicts.

hannes-ucsc commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,218 @@
+        """
+        self.in_ = kwargs['in']
+        self.json_validators = dict(jsonschema.Draft4Validator.VALIDATORS)
+        self.name = name
```
Replace this boiler-plate with `attr.s`

amarjandu replied:
Jesse and I spent some time looking at this, the issue we found was that the parameter `in` is used within the schema that defines the parameter: https://swagger.io/specification/#parameterObject . I'll take another look, not sure if it can be done. https://github.com/python-attrs/attrs/issues/381

hannes-ucsc commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,218 @@
+
+
+class SpecValidator:
+
```
docstring

hannes-ucsc commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,218 @@
+
+
+class SchemaSpecValidator(SpecValidator):
+
```
docstring

hannes-ucsc commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,218 @@
+        schema_type = self.schema['type']
+        if schema_type == 'string':
+            if self.schema.get('format') == 'uuid':
+                # noinspection PyUnusedLocal
```
prepend underscore to name of unused variable that also gets rid of the warning.

amarjandu replied:
this section was removed in favor of pattern specification for UUIDs

hannes-ucsc reviewed (changes requested):
I mostly got caught on cosmetics. You can start addressing those comment and answer my question(s), but I think you should also give me high-level walk-through of the mechanism in the parking lot.

hannes-ucsc commented:
Note the conflicts. No bump, since review was partial.

amarjandu committed:
fixup! [R 2/2] Validate requests using Swagger spec (#1465)

hannes-ucsc reviewed (commented):
![image](https://user-images.githubusercontent.com/5143256/106976090-a296ea80-670c-11eb-897b-030d1e9fa254.png)

Note that your branch is based on a merge commit but on a merged commit. I'll fix it.

hannes-ucsc commented on src/azul/chalice.py:
```diff
@@ -145,3 +153,107 @@ def _log_response(self, response):
+        endpoint.
+        """
+
+        azul_chalice_decorator = super().route(*args, **kwargs)
```
The smaller the [scope](https://en.wikipedia.org/wiki/Scope_(computer_science)) the shorter the variables can be.

```suggestion
        super_decorator = super().route(*args, **kwargs)
```

hannes-ucsc commented on src/azul/chalice.py:
```diff
@@ -145,3 +153,107 @@ def _log_response(self, response):
+                    status = self.validate_params(self.current_request)
+                    if isinstance(status, Response):
+                        return status
+                    return view_func(*args, **kwargs)
```
else

hannes-ucsc commented on src/azul/chalice.py:
```diff
@@ -145,3 +153,107 @@ def _log_response(self, response):
+        if validate:
+            def decorator(view_func):
+                @functools.wraps(view_func)
+                def f(*args, **kwargs):
```
```suggestion
                def wrapper(*args, **kwargs):
```

hannes-ucsc commented on src/azul/chalice.py:
```diff
@@ -145,3 +153,107 @@ def _log_response(self, response):
+            return azul_chalice_decorator
+
+    def get_spec_validators(self, path, http_method) -> Dict[str, SpecValidator]:
+        spec_validators = dict()
```
this can be a dict comprehension

hannes-ucsc commented on src/azul/chalice.py:
```diff
@@ -145,3 +153,107 @@ def _log_response(self, response):
+
+    def get_spec_validators(self, path, http_method) -> Dict[str, SpecValidator]:
+        spec_validators = dict()
+        for parameter_spec in [
```
use chain from itertools

hannes-ucsc commented on src/azul/chalice.py:
```diff
@@ -145,3 +153,107 @@ def _log_response(self, response):
+    def get_spec_validators(self, path, http_method) -> Dict[str, SpecValidator]:
+        spec_validators = dict()
+        for parameter_spec in [
+            *self.specs['paths'][path].get('parameters', []),
```
use empty tuple for default. It's should be cheaper. Tuples are immutable so they can be shared. In fact, there is only one empty tuple instance in Python. But most times you write `[]`, a new mutable list is allocated on the heap.

```
id([])
140129130225216
id([])
140129130206336
id([])
140129130206336
id([])
140129130206784
id([])
140129130206336
```

There seems to be some reuse going on. But the empty tuple is always the same

```
id(())
140129154482240
id(())
140129154482240
id(())
140129154482240
id(())
140129154482240
id(())
140129154482240
```

hannes-ucsc commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,242 @@
+    schema: Optional[JSON] = None
+    content: Optional[JSON] = None
+
+    def to_dict(self):
```
attrs has stuff for this

amarjandu replied:
looking into this `attr.fields_dict` does not play nicely with exceptions.

amarjandu replied:
I think using a custom to_dict method is advantageous here, it allows us to remove inapplicable fields for a given parameter. For example, we don't need `contents=None` for a parameter that defines a `schema`.

amarjandu replied:
ah! can use attr.asdict() and pass in a filter 👍🏽

hannes-ucsc commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,242 @@
+        :return: Dictionary of atttributes that are not None
+        """
+        return {
+            attribute.name: self.__getattribute__(attribute.name)
```
also you shouldn't call dunder methods directly unless there is no alternative.

hannes-ucsc commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,242 @@
+
+
+@attr.s(auto_attribs=True, kw_only=True)
+class SpecValidator(ABC):
```
Most of your attr.s instances are mutable. That worries me. Validators typically aren't mutable. Favor immutability,

hannes-ucsc commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,242 @@
+    A base class for OpenAPI specification validators. See
+    https://swagger.io/specification/#parameter-object for details about
+    parameter specification. The sub-classes for this class are broken down
+    by the complexity of the parameter that is being defined.
```
I do not understand "broken down by the complexity of the parameter that is being defined"

amarjandu replied:
Removed to avoid confusion

hannes-ucsc commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,242 @@
+    def validate(self, param_value: str):
+        """
+        Validates if a given request parameter meets the specification.
+        :param param_value: The value of the parameter to be validated
```
```suggestion
        
        :param param_value: The value of the parameter to be validated
```

Here and everywhere. This is a paragraph. We always use empty lines between paragraphs.

hannes-ucsc commented on src/azul/openapi/validation.py:
```diff
@@ -0,0 +1,242 @@
+@attr.s(auto_attribs=True, kw_only=True)
+class SchemaSpecValidator(SpecValidator):
+    """
+    SchemaSpecValidators are used to validate parameters that describe
```
Parameters don't describe schemas. The other way around, maybe. This problem seems to be recurring in the other docstrings.

hannes-ucsc reviewed (changes requested):
I think we should break this up into multiple PR. Think about which changes you can extract into separate stand-alone PRs. Luckily we can have multiple PRs per issue in Zenhub. It appears that the commits that tighten the spec fall into that category. There still be a big PR at the end but I want to pick some low hanging  fruit first.

amarjandu commented on lambdas/service/app.py:
```diff
@@ -991,23 +991,23 @@ def repository_search(entity_type: str, item_id: Optional[str]) -> JSON:
+            'oneOf': [
+                schema.object(is_=schema.array({})),
+                *(
+                    schema.object(**{
```
https://github.com/DataBiosphere/azul/pull/1849#pullrequestreview-546666351

TODO: try to switch back to the old version that uses object_type, and showcase the failure if it occurs. Why is `additionalProperties=False` for the inner-field introduced?

amarjandu replied:
```
Index: lambdas/service/app.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- lambdas/service/app.py	(revision 3928dce202978ca813ded725abf67a0b614bd7e3)
+++ lambdas/service/app.py	(date 1612806058041)
@@ -761,26 +761,26 @@
 
 filters_param_spec = params.query(
     'filters',
-    schema.optional(application_json(schema.object_type({
-        facet: {
-            'oneOf': [
-                schema.object(is_=schema.array({})),
-                *(
-                    schema.object(**{
-                        op: schema.array({},
-                                         items=schema.array({}, minItems=2, maxItems=2),
-                                         minItems=1,
-                                         maxItems=2)
-                    })
-                    for op in ['contains', 'within', 'intersects']
-                )
-            ]
-        }
-        for facet in app.facets
-    },
-        default='{}',
-        example={'cellCount': {'within': [[10000, 1000000000]]}},
-        additionalProperties=False
+    schema.optional(application_json(schema.object_type(
+        default='{}',
+        example={'cellCount': {'within': [[10000, 1000000000]]}},
+        properties={
+            facet: {
+                'oneOf': [
+                    schema.object(is_=schema.array({})),
+                    *(
+                        schema.object_type({
+                            op: schema.array(item=schema.array({}, minItems=2, maxItems=2),
+                                             minItems=1,
+                                             maxItems=2),
+
+                        })
+                        for op in ['contains', 'within', 'intersects']
+                    )
+                ]
+            }
+            for facet in app.facets
+        }
     ))),
     # FIXME: Spec for `filters` argument should be driven by field types
     #        https://github.com/DataBiosphere/azul/issues/2254
Index: src/azul/openapi/schema.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/azul/openapi/schema.py	(revision 3928dce202978ca813ded725abf67a0b614bd7e3)
+++ src/azul/openapi/schema.py	(date 1612805346578)
@@ -361,7 +361,7 @@
 }
 
 
-def object_type(properties: JSON, /, **kwargs) -> JSON:
+def object_type(properties: JSON, **kwargs) -> JSON:
     """
     Returns the schema for a JSON object with the given properties.
 
```

By not constraining w/ `additionalProperties=False` within the innter-field, invalid requests make it past the JSONSchema validator without raising an exception, There are downstream errors that occur from the viewfunction being called with invalid parameters. 
<img width="406" alt="Screen Shot 2021-02-08 at 10 15 57 AM" src="https://user-images.githubusercontent.com/1947534/107263638-d87eec00-69f6-11eb-94da-33a5c234981b.png">

amarjandu commented on src/azul/chalice.py:
```diff
@@ -145,3 +153,107 @@ def _log_response(self, response):
+    def __init__(self, app_name, unit_test=False, spec=None):
+        super().__init__(app_name=app_name, unit_test=unit_test, spec=spec)
+
+    # override so we can call _validate
```
removed `_validate`, which was being used as a wrapper for `validate_params`. Removing this comment.

hannes-ucsc commented:
No need to ask for review until the base PR has been approved.

amarjandu commented:
self note: IT passes

amarjandu commented on src/azul/chalice.py:
```diff
@@ -283,3 +297,112 @@ def _register_handler(self,
+        extra_params = all_provided_params.keys() - spec_parameters.keys()
+        if extra_params:
+            invalid_body['extra_parameters'] = sorted(extra_params)
+
```
Check to see if extra nested parameters work for accessions. It should work, but its dusty here.

amarjandu reviewed (commented):
Upkeep on direction for this.
